### PR TITLE
Graph break on differentiable boolean mask setitem

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102843

Fixes https://github.com/pytorch/pytorch/issues/102841

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy

### CODE REVIEW

File: torch/_dynamo/variables/tensor.py [hunk 1]
```diff
@@ -424,6 +424,18 @@ def call_method(
...
             return self.call_method(tx, "size", [ConstantVariable(0, **options)], {})
         elif name == "__setitem__":
+            if (
+                not (config.dynamic_shapes and config.capture_dynamic_output_shape_ops)
+                and any(
+                    isinstance(a, TensorVariable)
+                    and a.dtype in (torch.bool, torch.int8)
+                    for a in args
+                )
+                and any(isinstance(a, TensorVariable) and a.requires_grad for a in args)
```
Comment (voznesenskym) [inline torch/_dynamo/variables/tensor.py:434]: You loop over args twice, and do isinstance(a, TensorVariable) twice.

```
any(
    isinstance(a, TensorVariable)
    and (a.dtype in (torch.bool, torch.int8))
    and a.requires_grad
    for a in args
)
```

File: torch/_dynamo/variables/tensor.py [hunk 2]
```diff
@@ -424,6 +424,18 @@ def call_method(
         elif name == "__len__":
             return self.call_method(tx, "size", [ConstantVariable(0, **options)], {})
         elif name == "__setitem__":
+            if (
+                not (config.dynamic_shapes and config.capture_dynamic_output_shape_ops)
```
Comment (voznesenskym) [inline torch/_dynamo/variables/tensor.py:428]: This is sliding us a bit back on the dynamic shapes checks, but, since I am all in on FSDP atm, this is more than reasonable.

Do you mind just adding a comment as to why this is excluded like this?

**Review by voznesenskym — APPROVED** [review]

File: torch/_dynamo/variables/tensor.py (hunk 2 above)
Comment (ezyang) [reply torch/_dynamo/variables/tensor.py:428]: This is just following the same pattern we do for nonzero. When you kill config.dynamic_shapes you can just delete that from here and the knob is now just `not capture_dynamic_output_shape_ops`, which makes sense (because inductor can't deal with nonzero right now).
